Tests/Foundation/Tests/TestNSError.swift

```diff
@@ -106,6 +111,59 @@ class TestNSError : XCTestCase {
             XCTFail()
         }
     }
+    
+    func test_ConvertErrorToNSError_domain() {
+        struct CustomSwiftError: Error {
+        }
+        XCTAssertTrue((CustomSwiftError() as NSError).domain.contains("CustomSwiftError"))
+    }
+    
+    func test_ConvertErrorToNSError_errorCode() {
+        enum SwiftError: Error {
+            case zero
+            case one
+            case two
+        }
+        
+        XCTAssertEqual((SwiftError.zero as NSError).code, 0)
+        XCTAssertEqual((SwiftError.one as NSError).code, 1)
+        XCTAssertEqual((SwiftError.two as NSError).code, 2)
+    }
+    
+    func test_ConvertErrorToNSError_errorCodeRawInt() {
+        enum SwiftError: Int, Error {
+            case minusOne = -1
+            case fortyTwo = 42
+        }
+        
+        XCTAssertEqual((SwiftError.minusOne as NSError).code, -1)
+        XCTAssertEqual((SwiftError.fortyTwo as NSError).code, 42)
+    }
+    
+    func test_ConvertErrorToNSError_errorCodeRawUInt() {
+        enum SwiftError: UInt, Error {
+            case fortyTwo = 42
+        }
+        
+        XCTAssertEqual((SwiftError.fortyTwo as NSError).code, 42)
+    }
+    
+    func test_ConvertErrorToNSError_errorCodeWithAssosiatedValue() {
+        enum SwiftError: Error {
+            case one
+            case two
+            case three(String)
+            case four
+            case five(String)
+        }
+        
+        XCTAssertEqual((SwiftError.one as NSError).code, 2)
+        XCTAssertEqual((SwiftError.two as NSError).code, 3)
+        XCTAssertEqual((SwiftError.three("three") as NSError).code, 0)
+        XCTAssertEqual((SwiftError.four as NSError).code, 4)
+        XCTAssertEqual((SwiftError.five("five") as NSError).code, 1)
```

Comment:
It's not quite clear why these values are what they are, especially for those with associated values. Could you add a comment to the test that clarifies this please?

Reply:
The test was initially based on actual observed behavior on Darwin platform. In search of details I've found that default error code for enum case is based on enum tag. Tag generation logic is located in EnumImplStrategy::getTagIndex ( https://github.com/apple/swift/blob/d6227dcca01b74b64f078949e69e91324e39f815/lib/IRGen/GenEnum.cpp#L239-L255 ). It starts from cases with payload (.three and .five), and then proceeds to cases without payload (.one , .two , .four ). This is why we are observing such error codes: .three = 0, .five = 1, .one = 2, .two = 3, .four = 4.

Reply:
I didn't know that, very interesting behavior. Glad that it could become consistent across platforms.